docs/proposals/failover/application-failover.md

```diff
@@ -0,0 +1,566 @@
+    // If specified, all gates will be evaluated for prerequisites.
+    // +optional
+    InitiateGates *FailoverInitiateGate `json:"initiateGates,omitempty"`
+    
+    // ActiveGates indicates the conditions of performing the failover process.
+    //
+    // +kubebuilder:validation:MinItems=1
+    // +required
+    ActiveGates *FailoverActiveGate `json:"activeGates"`
+    
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Immediately", "Graciously" and "Never".
+    // Defaults to "Graciously".
+    //
+    // +kubebuilder:validation:Enum=Immediately;Graciously;Never
+    // +optional
+    PurgeMode string `json:"purgeMode,omitempty"`
+    
+    // EscapeSeconds represents the period of time the cluster from which the
+    // application was migrated from should be escaped.
+    // Defaults to 600s.
+    //
+    // +kubebuilder:default=600
+    // +optional
+    EscapeSeconds int32 `json:"escapeSeconds,omitempty"`
+}
+
+
+// FailoverInitiateGate represents the preconditions of the failover process.
+type FailoverInitiateGate struct {
+    // DelaySeconds refers to a period of time after the application has started before
+	// the application failover is initiated.
+    // +optional
+    DelaySeconds int32 `json:"delaySeconds,omitempty"`
+    
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Healthy".
+    // +optional
+    HealthyState string `json:"healthyState,omitempty"`
+}
+
+// FailoverActiveGate represents the conditions of performing the failover process.
+type FailoverActiveGate struct {
+    // TolerationSeconds represents the period of time the Karmada should wait
+    // before performing failover. Defaults to 10s.
+    //
+    // +kubebuilder:default=10
+    // +optional
+    TolerationSeconds int32
+
+    // HealthyState refers to the healthy status reported by the Karmada resource
+    // interpreter.
+    // Valid options are "Unhealthy" and "Unknown".
+    // +optional
+    HealthyState []string `json:"healthyState,omitempty"`
+}
+```
+
+For example:
+
+```yaml
+apiVersion: policy.karmada.io/v1alpha1
+kind: PropagationPolicy
+metadata:
+  name: nginx-propagation
+spec:
+  resourceSelectors:
+    - apiVersion: apps/v1
+      kind: Deployment
+      name: nginx
+    - apiVersion: apps/v1
+      kind: StatefulSet
+      name: mysql
+  failover:
+    application:
+      initiateGates:
+        delaySeconds: 5000
+      activeGates:
+        healthyState: 
+        - Unhealthy
+  placement:
+    clusterAffinity:
+      clusterNames:
+        - member1
+        - member2
+```
+
+#### ResourceBinding API change
+
+```golang
+type ResourceBindingSpec struct {
+    ...
+    // EvictedClusters represents the evicted clusters.
+    // +optional
+    EvictedClusters []EvictedCluster `json:"evictedClusters,omitempty"`
```

Comment:
I don't think this field should be put in `spec`, because it's not a declaration, actually, it is a kind of intermediate that is used by controllers(scheduler).

Reply:
I think it's a bit like `GracefulEvictionTasks`. It may also has a scene. The user manually fixed the failure within the recovery time set by the cluster. Users may manually make a cluster schedulable.
BTW, I have also considered using the taint mechanism to achieve this function, but it will introduce `cluster` object, which may make the function more complicated.

Reply:
After I thought about it, I'd like to pick the name `EvictionHistory` or `ActiveEvictionHistory`.

Reply:
What is the difference between this field and `GracefulEvictionTasks`?